multiplexer/abci/abci.go

```diff
@@ -0,0 +1,157 @@
+	}
+
+	resp, err := app.FinalizeBlock(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to finalize block: %w", err)
+	}
+
+	// set the app version to be used in the next block.
+	if resp.ConsensusParamUpdates != nil && resp.ConsensusParamUpdates.GetVersion() != nil {
+		m.nextAppVersion = resp.ConsensusParamUpdates.GetVersion().App
+	}
+
+	return resp, err
+}
+
+func (m *Multiplexer) Info(_ context.Context, req *abci.RequestInfo) (*abci.ResponseInfo, error) {
+	app, err := m.getApp()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get app for version %d: %w", m.appVersion, err)
+	}
+
+	return app.Info(req)
+}
+
+func (m *Multiplexer) InitChain(_ context.Context, req *abci.RequestInitChain) (*abci.ResponseInitChain, error) {
+	app, err := m.getApp()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get app for genesis: %w", err)
+	}
+	return app.InitChain(req)
+}
+
+func (m *Multiplexer) ListSnapshots(_ context.Context, req *abci.RequestListSnapshots) (*abci.ResponseListSnapshots, error) {
+	app, err := m.getApp()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get app for version %d: %w", m.appVersion, err)
+	}
+	return app.ListSnapshots(req)
+}
+
+func (m *Multiplexer) LoadSnapshotChunk(_ context.Context, req *abci.RequestLoadSnapshotChunk) (*abci.ResponseLoadSnapshotChunk, error) {
+	app, err := m.getApp()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get app for version %d: %w", m.appVersion, err)
+	}
+	return app.LoadSnapshotChunk(req)
+}
+
+func (m *Multiplexer) OfferSnapshot(_ context.Context, req *abci.RequestOfferSnapshot) (*abci.ResponseOfferSnapshot, error) {
+	app, err := m.getApp()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get app for version %d: %w", m.appVersion, err)
+	}
+	return app.OfferSnapshot(req)
+}
+
+func (m *Multiplexer) PrepareProposal(_ context.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+	app, err := m.getApp()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get app for version %d: %w", m.appVersion, err)
+	}
+	return app.PrepareProposal(req)
+}
+
+func (m *Multiplexer) ProcessProposal(_ context.Context, req *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
+	app, err := m.getApp()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get app for version %d: %w", m.appVersion, err)
+	}
+	return app.ProcessProposal(req)
+}
+
+func (m *Multiplexer) Query(ctx context.Context, req *abci.RequestQuery) (*abci.ResponseQuery, error) {
+	app, err := m.getApp()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get app for version %d: %w", m.appVersion, err)
+	}
+	return app.Query(ctx, req)
+}
+
+func (m *Multiplexer) VerifyVoteExtension(_ context.Context, req *abci.RequestVerifyVoteExtension) (*abci.ResponseVerifyVoteExtension, error) {
+	app, err := m.getApp()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get app for version %d: %w", m.appVersion, err)
+	}
+	return app.VerifyVoteExtension(req)
+}
```

Comment:
similar question to https://github.com/celestiaorg/celestia-app/pull/4553/files#r2042761731

I think I'm misunderstanding something, as it seems like we shouldn't be supporting verify vote extension for tendermint

Reply:
Maybe I need to add a diagram. Basically the multiplexer interacts with comet always using ABCI 2.0 (as we used comet 0.38). It would have done the reverse if we'd use 0.34.

Basically we have comet 0.38 (ABCI 2.0) -> multiplexer (remote v1 / v2) -> app v4 (v2)
                                                                                                                               -> app v3 (v1)
                                                                                                                                -> ...

Reply:
> the multiplexer interacts with comet always using ABCI 2.0

ahh ok I see thanks!